### PR TITLE
[Torch FX] Post Quantize Weights Compression

### PR DESCRIPTION
**Pull request by anzr299**
### Changes

Transformation for removing fake quantize nodes and saving all weights to disk in int8 format after quantization. It works as follows:
1. Reshape the scale if qdq operation is per-channel.
2. Pattern match the quantize-dequantize nodes.
3. Filter the matches to only include quantize-dequantize ops with constant input.
4. Replace with the multiplication of the scale and input.

### Reason for changes

To compress the model after quantization

### Tests

Add `test_post_quantization_compression()` in `tests/torch/fx/test_model_transformer.py` which checks the data type of all weights in the model after applying quantization and also checks the value after the decompression step (element-wise multiplication operation).

### Tickets
#2766

### CODE REVIEW

File: nncf/experimental/torch/fx/transformations.py [hunk 1]
```diff
@@ -528,6 +542,128 @@ def fuse_conv_bn(model: torch.fx.GraphModule) -> None:
...
+    scale_value = get_tensor_constant_from_node(scale_node, model)
+    weight_value = get_tensor_constant_from_node(weight_node, model)
+    axis = node.args[3]
+    new_shape = [1] * weight_value.dim()
+    new_shape[axis] = scale_value.shape[0]
+    scale_value = scale_value.reshape(new_shape)
+    constant_update_fn(model, node, scale_value, 1, updated_node_name=scale_node.name + "_updated_constant")
+
+
+def fq_weights_transformation(model: torch.fx.GraphModule) -> None:
```
Comment (alexsu52) [inline nncf/experimental/torch/fx/transformations.py:642]: I could not find a unit test for `fq_weights_transformation`.

File: nncf/experimental/torch/fx/transformations.py [hunk 2]
```diff
@@ -528,6 +542,128 @@ def fuse_conv_bn(model: torch.fx.GraphModule) -> None:
...
+                constant_update_fn(
+                    model,
+                    node.all_input_nodes[0],
+                    result,
+                    port_id,
+                    updated_node_name="compressed_weight_updated_constant",
+                )
+
+
+def compress_post_quantize_transformation(model: torch.fx.GraphModule):
```
Comment (alexsu52) [inline nncf/experimental/torch/fx/transformations.py:662]: I could not find a unit test for `compress_post_quantize_transformation`.

**Review by alexsu52 — COMMENTED** [review]
I saw that you updated the tests for all quantization pipelines and verified that the transform was applied, but you didn't added tests for corner cases. I would suggest adding unit tests to check for such cases.

File: nncf/experimental/torch/fx/transformations.py [hunk 3]
```diff
@@ -528,6 +542,128 @@ def fuse_conv_bn(model: torch.fx.GraphModule) -> None:
...
+
+
+def fq_weights_transformation(model: torch.fx.GraphModule) -> None:
+    for node in model.graph.nodes:
+        if node.target in [
+            torch.ops.quantized_decomposed.dequantize_per_tensor.default,
+            torch.ops.quantized_decomposed.dequantize_per_channel.default,
+        ]:
+            port_id = 0
+            input_tup = _get_node_inputs(node.all_input_nodes[0], model)
```
Comment (alexsu52) [inline nncf/experimental/torch/fx/transformations.py:649]: I would suggest to check that `node.all_input_nodes[0]` is            `torch.ops.quantized_decomposed.quantize_per_tensor.default` or      `torch.ops.quantized_decomposed.quantize_per_channel.default`

File: nncf/experimental/torch/fx/transformations.py [hunk 4]
```diff
@@ -528,6 +542,128 @@ def fuse_conv_bn(model: torch.fx.GraphModule) -> None:
     model.recompile()
 
 
+def _get_pattern_replacement_per_channel():
+    def pattern_per_channel(weight, scale, zero_point, axis, low, high, dtype):
```
Comment (alexsu52) [inline nncf/experimental/torch/fx/transformations.py:546]: I see you support for zero_point == 0. Could you add support for asymmetric quantization as well?

File: nncf/experimental/torch/fx/transformations.py (hunk 3 above)
Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py]: Done!

File: nncf/experimental/torch/fx/transformations.py [hunk 5]
```diff
@@ -528,6 +560,188 @@ def fuse_conv_bn(model: torch.fx.GraphModule) -> None:
...
+    input_tup = []
+    for params in node.args:
+        if isinstance(params, torch.fx.Node):
+            if params.op == "get_attr":
+                value = get_tensor_constant_from_node(params, model)
+                input_tup.append(value)
+            else:
+                return None
+        else:
+            input_tup.append(params)
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch/fx/transformations.py:668]:
```suggestion
    for arg in node.args:
        if isinstance(arg, torch.fx.Node):
            if arg.op != "get_attr":
                return None
            arg = get_tensor_constant_from_node(arg, model)
        input_tup.append(arg)
```

File: nncf/experimental/torch/fx/transformations.py [hunk 6]
```diff
@@ -528,6 +560,188 @@ def fuse_conv_bn(model: torch.fx.GraphModule) -> None:
...
+            if input_tup:
+                input_tup[0] = quantize_node.target(*tuple(input_tup))
+                result = node.target(*tuple(input_tup))
+                constant_update_fn(
+                    model,
+                    quantize_node,
+                    result,
+                    port_id,
+                    updated_node_name="compressed_weight_updated_constant",
+                )
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch/fx/transformations.py:737]:
```suggestion
        if node.target not in DEQUANTIZE_NODES:
            continue
        port_id = 0
        quantize_node = node.all_input_nodes[0]
        # Case where Quantize Dequantize nodes are not consecutive
        if quantize_node.target != QDQ_PAIR[node.target]:
            continue
        input_tup = _get_node_inputs(quantize_node, model)
        if not input_tup:
            continue
        input_tup[0] = quantize_node.target(*tuple(input_tup))
        result = node.target(*tuple(input_tup))
        constant_update_fn(
                model,
                quantize_node,
                result,
                port_id,
                updated_node_name="compressed_weight_updated_constant",
            )
```

File: nncf/experimental/torch/fx/transformations.py [hunk 7]
```diff
@@ -528,6 +560,188 @@ def fuse_conv_bn(model: torch.fx.GraphModule) -> None:
...
+    return matches_per_channel, matches_per_tensor
+
+
+def _get_node_inputs(node: torch.fx.Node, model: torch.fx.GraphModule) -> Optional[Tuple[Union[torch.Tensor, int]]]:
+    """
+    Gets the inputs for the Quantize node which quantize the weights. Otherwise return None
+    :param node: Node to get the inputs from
+    :param model: Model to which the node belongs
+    """
+    input_tup = []
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch/fx/transformations.py:659]: What do you think about renaming `input_tup` to `args` here and elsewhere?

File: nncf/experimental/torch/fx/transformations.py [hunk 8]
```diff
@@ -528,6 +560,188 @@ def fuse_conv_bn(model: torch.fx.GraphModule) -> None:
...
+def _compress_qdq_constant_transformation(model: torch.fx.GraphModule) -> None:
+    """
+    Change the FP32 weight value to Int8 and also reshape the scale for per_channel_quantization.
+    :param: model: Model to apply transformations to.
+    """
+    for node in model.graph.nodes:
+        if node.target in DEQUANTIZE_NODES:
+            quantize_node = node.all_input_nodes[0]
+            if quantize_node.target != QDQ_PAIR[node.target]:
+                pass
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch/fx/transformations.py:681]: Please remove this if

File: nncf/experimental/torch/fx/transformations.py [hunk 9]
```diff
@@ -528,6 +560,188 @@ def fuse_conv_bn(model: torch.fx.GraphModule) -> None:
...
+    return input_tup
+
+
+def _compress_qdq_constant_transformation(model: torch.fx.GraphModule) -> None:
+    """
+    Change the FP32 weight value to Int8 and also reshape the scale for per_channel_quantization.
+    :param: model: Model to apply transformations to.
+    """
+    for node in model.graph.nodes:
+        if node.target in DEQUANTIZE_NODES:
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch/fx/transformations.py:678]: Let's decrease amount of indents here
```suggestion
        if node.target not in DEQUANTIZE_NODES:
            continue
        ....
```

File: tests/torch/fx/test_model_transformer.py [hunk 10]
```diff
@@ -136,6 +152,78 @@ def test_output_insertion_transformation(tuple_output, target_point):
...
+        input_ids = torch.ones(model_case.input_shape)
+        exported_model = capture_pre_autograd_graph(model, args=(input_ids,))
+        quantization_parameters["advanced_parameters"] = nncf.AdvancedQuantizationParameters()
+        quantization_parameters["subset_size"] = 1
+        quantized_model = nncf.quantize(
+            exported_model, calibration_dataset=nncf.Dataset([input_ids]), **quantization_parameters
+        )
+
+    for node in quantized_model.graph.nodes:
+        if node.name[:3] == "mul":
```
Comment (daniil-lyakhov) [inline tests/torch/fx/test_model_transformer.py:184]: we can reduce amount of indents here as well

File: tests/torch/fx/test_model_transformer.py [hunk 11]
```diff
@@ -136,6 +152,78 @@ def test_output_insertion_transformation(tuple_output, target_point):
...
+
+    for node in quantized_model.graph.nodes:
+        if node.target in [
+            torch.ops.quantized_decomposed.dequantize_per_channel.default,
+            torch.ops.quantized_decomposed.dequantize_per_tensor.default,
+        ]:
+            input_tup = []
+            quantize_node = node.args[0]
+            input_tup = _get_node_inputs(quantize_node, quantized_model)
+            if input_tup:
```
Comment (daniil-lyakhov) [inline tests/torch/fx/test_model_transformer.py:219]: you should check amount of nodes here as well. It is possible that all `input_tup` will be None for some reason but test will be green

File: tests/torch/fx/test_model_transformer.py [hunk 12]
```diff
@@ -136,6 +152,78 @@ def test_output_insertion_transformation(tuple_output, target_point):
...
+    model = getattr(models, model_id)
+    return ModelCase(partial(model, weights=None), model_id, input_shape)
+
+
+TEST_MODELS_QUANIZED = (
+    (torchvision_model_case("resnet18", (1, 3, 224, 224)), {}),
+    (torchvision_model_case("mobilenet_v3_small", (1, 3, 224, 224)), {}),
+    (torchvision_model_case("vit_b_16", (1, 3, 224, 224)), {"model_type": nncf.ModelType.TRANSFORMER}),
+    (torchvision_model_case("swin_v2_s", (1, 3, 224, 224)), {"model_type": nncf.ModelType.TRANSFORMER}),
+)
```
Comment (daniil-lyakhov) [inline tests/torch/fx/test_model_transformer.py:165]: Maybe it worth to test only one synthetic model here? This will increase time significantly ( there will be x3 model quantizations than before). Besides, new supported models should be specified in `test_models` and here, and this approach looks ambiguous
CC: @alexsu52

File: tests/torch/fx/test_model_transformer.py [hunk 13]
```diff
@@ -136,6 +152,78 @@ def test_output_insertion_transformation(tuple_output, target_point):
...
+        quantization_parameters["subset_size"] = 1
+        quantized_model = nncf.quantize(
+            exported_model, calibration_dataset=nncf.Dataset([input_ids]), **quantization_parameters
+        )
+
+    for node in quantized_model.graph.nodes:
+        if node.name[:3] == "mul":
+            input_tup = []
+            input_tup = _get_node_inputs(node, quantized_model)
+            if input_tup:
```
Comment (daniil-lyakhov) [inline tests/torch/fx/test_model_transformer.py:187]: What if all `input_tup` is None? Test will be green anyway. Please check amount of nodes here

**Review by daniil-lyakhov — CHANGES_REQUESTED** [review]

File: nncf/experimental/torch/fx/transformations.py [hunk 14]
```diff
@@ -27,9 +28,26 @@
...
+    torch.ops.quantized_decomposed.quantize_per_channel.default,
+]
+DEQUANTIZE_NODES = [
+    torch.ops.quantized_decomposed.dequantize_per_tensor.default,
+    torch.ops.quantized_decomposed.dequantize_per_channel.default,
+]
+# Map quantize_per_tensor to dequantize_per_tensor, the same for per_channel and vice-versa
+QDQ_PAIR = {
+    **{quantize_node: dequantize_node for quantize_node, dequantize_node in zip(QUANTIZE_NODES, DEQUANTIZE_NODES)},
+    **{dequantize_node: quantize_node for quantize_node, dequantize_node in zip(QUANTIZE_NODES, DEQUANTIZE_NODES)},
```
[the remark this answers is not in the document]
Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py:42]: I created a mapping between the quantize node and its pair of dequantize node and vice versa. So something like this:
quantize_per_channel: dequantize_per_channel
dequantize_per_channel: dequantize_per_channel
and so on.

File: tests/torch/fx/test_model_transformer.py (hunk 12 above)
Comment (anzr299) [reply tests/torch/fx/test_model_transformer.py:165]: Alright, I could test only a resnet model to save time.

Comment (daniil-lyakhov) [reply tests/torch/fx/test_model_transformer.py:165]: Does it covers all the cases?

File: nncf/experimental/torch/fx/transformations.py [hunk 15]
```diff
@@ -39,18 +57,23 @@ def _set_new_node_meta(
...
-        prev_node.meta["val"]
-        if prev_node.op not in ["get_attr"]
-        else get_tensor_constant_from_node(prev_node, model).data
-    )
-    val = val if isinstance(val, tuple) else (val,)
-    retval = []
-    for t in val:
-        retval.append(torch.ones(t.shape))
+    vals = []
+    prev_nodes = [prev_nodes] if not isinstance(prev_nodes, Iterable) else prev_nodes
```
[the remark this answers is not in the document]
Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py:61]: Do you mean passing in only a tuple to the function?

[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/experimental/torch/fx/transformations.py:61]: Yes

File: nncf/experimental/torch/fx/transformations.py (hunk 5 above)
Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py:668]: Yeah, this looks better. I will change it!

File: nncf/experimental/torch/fx/transformations.py (hunk 7 above)
Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py:659]: I assumed it could cause some confusion with node.args or other args variables too. Although I think it will be better since it returns the args but with tensors. I will change it

File: nncf/experimental/torch/fx/transformations.py [hunk 16]
```diff
@@ -528,6 +560,188 @@ def fuse_conv_bn(model: torch.fx.GraphModule) -> None:
...
+    for params in node.args:
+        if isinstance(params, torch.fx.Node):
+            if params.op == "get_attr":
+                value = get_tensor_constant_from_node(params, model)
+                input_tup.append(value)
+            else:
+                return None
+        else:
+            input_tup.append(params)
+    return input_tup
```
[the remark this answers is not in the document]
Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py:669]: Done

File: tests/torch/fx/test_model_transformer.py (hunk 12 above)
Comment (anzr299) [reply tests/torch/fx/test_model_transformer.py:165]: it covers most of the cases except `quantize_per_tensor` node following a constant node.

File: nncf/experimental/torch/fx/transformations.py [hunk 17]
```diff
@@ -528,6 +560,189 @@ def fuse_conv_bn(model: torch.fx.GraphModule) -> None:
     model.recompile()
 
 
+def _get_pattern_replacement_per_channel() -> None:
```
[the remark this answers is not in the document]
Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py:563]: Done

File: nncf/experimental/torch/fx/transformations.py (hunk 8 above)
Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py:681]: That was a mistake and was supposed to be a continue instead. I have changed it now :+1:

File: tests/torch/fx/test_model_transformer.py (hunk 12 above)
Comment (anzr299) [reply tests/torch/fx/test_model_transformer.py:165]: > Maybe it worth to integrate this tests into `test_models`? So, after the model is being checked by NNCFGraph we are checking its compressed weights?

Since this is a transformation test and checks if the backend parameters are working correctly too, I had decided to put it in `model_transformer_test`. Should I move it to `test_models`?

File: tests/torch/fx/test_model_transformer.py (hunk 13 above)
Comment (anzr299) [reply tests/torch/fx/test_model_transformer.py:187]: Done!

File: tests/torch/fx/test_model_transformer.py (hunk 11 above)
Comment (anzr299) [reply tests/torch/fx/test_model_transformer.py:219]: Done!

File: nncf/experimental/torch/fx/transformations.py (hunk 7 above)
Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py:659]: Done

File: nncf/experimental/torch/fx/transformations.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py:61]: Alright, Done :+1:

File: nncf/experimental/torch/fx/transformations.py [hunk 18]
```diff
@@ -39,18 +57,23 @@ def _set_new_node_meta(
...
+                prev_node.meta["val"]
+                if prev_node.op not in ["get_attr"]
+                else get_tensor_constant_from_node(prev_node, model).data
+            )
+            if isinstance(prev_node, torch.fx.Node)
+            else prev_node
+        )
 
+        val = val[0] if isinstance(val, tuple) else val
+        vals.append(val)
```
[the remark this answers is not in the document]
Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py:74]: Alright, Done.

File: tests/torch/fx/test_model_transformer.py (hunk 12 above)
Comment (anzr299) [reply tests/torch/fx/test_model_transformer.py]: It looks a little redundant, I will move it to `test_models` inside of `test_quantized_model`

File: nncf/experimental/torch/fx/quantization/quantize_model.py [hunk 19]
```diff
@@ -108,6 +116,8 @@ def quantize_impl(
     quantized_model.meta.update(original_graph_meta)
     quantized_model = _disallow_eval_train(quantized_model)
 
+    quantized_model = GraphModule(quantized_model, quantized_model.graph)
+
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch/fx/quantization/quantize_model.py:120]: Could you please add a comment here describing why we need to create a new `GraphModule`?

File: nncf/experimental/torch/fx/transformations.py [hunk 20]
```diff
@@ -182,7 +220,13 @@ def constant_update_transformation(model: torch.fx.GraphModule):
...
 
 
-def constant_update_fn(model: torch.fx.GraphModule, node: torch.fx.Node, value: torch.Tensor, input_port_id: int = 1):
+def constant_update_fn(
+    model: torch.fx.GraphModule,
+    node: torch.fx.Node,
+    value: torch.Tensor,
+    input_port_id: int = 1,
+    updated_node_name: Optional[str] = None,
+):
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch/fx/transformations.py:229]: Docstring is missed

File: nncf/experimental/torch/fx/quantization/quantize_model.py (hunk 19 above)
Comment (anzr299) [reply nncf/experimental/torch/fx/quantization/quantize_model.py:120]: Alright

File: nncf/experimental/torch/fx/transformations.py (hunk 20 above)
Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py:229]: Alright, Done!

File: nncf/experimental/torch/fx/quantization/quantize_model.py (hunk 19 above)
Comment (anzr299) [reply nncf/experimental/torch/fx/quantization/quantize_model.py:120]: Done

File: nncf/experimental/torch/fx/transformations.py [hunk 21]
```diff
@@ -593,7 +837,7 @@ def separate_linear_and_bias(model: torch.fx.GraphModule):
     for n in model.graph.nodes:
         if not _is_linear(n):
             continue
-        if len(n.args) < 3 or n.args[2] is None:
+        if len(n.args) < 3 or n.args[2] is None or n.args[1].op != "get_attr":
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch/fx/transformations.py:840]: What's the purpose of this check?

File: nncf/experimental/torch/fx/transformations.py [hunk 22]
```diff
@@ -636,7 +880,7 @@ def separate_conv_and_bias(model: torch.fx.GraphModule):
     for n in model.graph.nodes:
         if not _is_conv(n):
             continue
-        if len(n.args) < 3 or n.args[2] is None:
+        if len(n.args) < 3 or n.args[2] is None or n.args[1].op != "get_attr":
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch/fx/transformations.py:883]: What's the purpose of this check?

Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py:883]: This is to ensure that the weight node is a constant. It was giving error in the case of a model with already inserted quantize-dequantize node for example:
weight   bias activation
    |           |        |
quantize  |        |   
    |           |        |
  add        |        |
    |           \|       |
dequantize \     |
     \             |    / 
         Conv2d
         
If the following pattern is passed, the convolution node does not have any check to ignore such a pattern.

File: nncf/experimental/torch/fx/transformations.py (hunk 21 above)
Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py:840]: https://github.com/openvinotoolkit/nncf/pull/2984#discussion_r1802694964

File: nncf/experimental/torch/fx/transformations.py (hunk 22 above)
Comment (daniil-lyakhov) [reply nncf/experimental/torch/fx/transformations.py:883]: Please add a comment with description. From my understanding, it's not a problem of the approach but rather a not supported case.
Second question: where does this occur? I believe that quantized models should not be quantized twice

Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py:885]: It is an edge cases as @alexsu52 suggested. So, this minor change could help support such a case. I will add a comment

File: nncf/experimental/torch/fx/transformations.py [hunk 23]
```diff
@@ -39,18 +79,22 @@ def _set_new_node_meta(
...
-        if prev_node.op not in ["get_attr"]
-        else get_tensor_constant_from_node(prev_node, model).data
-    )
-    val = val if isinstance(val, tuple) else (val,)
-    retval = []
-    for t in val:
-        retval.append(torch.ones(t.shape))
-
+    vals = []
+    prev_nodes = [prev_nodes] if not isinstance(prev_nodes, Iterable) else prev_nodes
```
[the remark this answers is not in the document]
Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py]: `module_insertion_transformation` calls this function with a single node, which is why I added this. Would you suggest I modify the call to pass a list of the single node?
Like so:
`_set_new_node_meta(new_node, target_node, module_to_insert, model)`
to:
`_set_new_node_meta(new_node, [target_node], module_to_insert, model)`

[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply nncf/experimental/torch/fx/transformations.py]: `_set_new_node_meta` takes `prev_nodes: Tuple[Argument, ...],`, so this code is redundant. I saw you have removed this line, thanks!

File: tests/torch/fx/test_models.py [hunk 24]
```diff
@@ -161,16 +194,51 @@ def transform_fn(data_item):
...
     # Uncomment to visualize torch fx graph
     # from tests.torch.fx.helpers import visualize_fx_model
     # visualize_fx_model(quantized_model, f"{model_case.model_id}_int8.svg")
-
+    save_dir = FX_QUANTIZED_COMPRESSED_DIR_NAME if compress_weights else FX_QUANTIZED_DIR_NAME
     nncf_graph = GraphConverter.create_nncf_graph(quantized_model)
-    check_graph(nncf_graph, get_dot_filename(model_case.model_id), FX_QUANTIZED_DIR_NAME, extended=True)
+    check_graph(nncf_graph, get_dot_filename(model_case.model_id), save_dir, extended=True)
+    q_nodes, dq_nodes = count_q_dq(quantized_model)
+    print("THIS: ", q_nodes, dq_nodes)
```
Comment (daniil-lyakhov) [inline tests/torch/fx/test_models.py:210]:
```suggestion
```

**Review by daniil-lyakhov — APPROVED** [review]
LGTM

File: tests/torch/fx/test_models.py (hunk 24 above)
Comment (anzr299) [reply tests/torch/fx/test_models.py]: 🤦

File: nncf/experimental/torch/fx/transformations.py [hunk 25]
```diff
@@ -529,9 +582,206 @@ def fuse_conv_bn(model: torch.fx.GraphModule) -> None:
...
+
+
+def compress_post_quantize_transformation(model: torch.fx.GraphModule) -> None:
+    """
+    Applies transformation to compress the weights to Int8 after the quantization step.
+
+    :param model: Model to apply transformations to.
+    """
+    _compress_qdq_constant_transformation(model)
+    _remove_constant_qdq_transformation(model)
```
Comment (alexsu52) [inline nncf/experimental/torch/fx/transformations.py:778]: It looks like `_compress_qdq_constant_transformation` returns incorrect model. I mean the model could not be infered. I suggest to simplify structure of this function and do all changes in one transformation which replace the FP32->q->dq pattern to the INT8-> sub -> mul pattern. Inside of the pattern you can calculate INT8 constant and reshape scale and zero_point using util functions.

Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py:778]: Alright, this seems more robust. I will do it 👍

Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py:792]: Done

File: nncf/experimental/torch/fx/transformations.py [hunk 26]
```diff
@@ -529,9 +583,217 @@ def fuse_conv_bn(model: torch.fx.GraphModule) -> None:
...
+    scale = scale.reshape(new_shape)
+    zp = zp.reshape(new_shape)
+    constant_update_fn(model, sub_node, zp, 1, updated_node_name="zero_point_updated_constant")
+    constant_update_fn(model, mul_node, scale, 1, updated_node_name="scale_updated_constant")
+
+
+def fq_weights_transformation(model: torch.fx.GraphModule) -> None:
+    """
+    This function applies a transformation to replace the FP32 weights with Fake Quantized
+    FP values to avoid the rounding error when converting to OpenVino model.
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch/fx/transformations.py:759]: Unused function

File: nncf/experimental/torch/fx/transformations.py [hunk 27]
```diff
@@ -529,9 +583,217 @@ def fuse_conv_bn(model: torch.fx.GraphModule) -> None:
...
+        scale_node = get_const(nodes_map["scale"])
+        zp_node = get_const(nodes_map["zero_point"])
+        axis = nodes_map["axis"]
+        port_id = 0
+        result = None
+        if axis is not None:
+            result = torch.ops.quantized_decomposed.quantize_per_channel.default(
+                weight_node, scale_node, zp_node, axis, -128, 127, torch.int8
+            )
+            _reshape_scale_zp(model, sub_node, mul_node, weight_node, scale_node, zp_node, axis)
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch/fx/transformations.py:722]: Let's keep qdq creation in scope of one if else statement
```suggestion
        if axis is not None:
            result = torch.ops.quantized_decomposed.quantize_per_channel.default(
                weight_node, scale_node, zp_node, axis, -128, 127, torch.int8
            )
            _reshape_scale_zp(model, sub_node, mul_node, weight_node, scale_node, zp_node, axis)
        else:
            torch.ops.quantized_decomposed.quantize_per_tensor.default(
                weight_node, scale_node, zp_node, -128, 127, torch.int8
            )
            ...
```

File: nncf/experimental/torch/fx/transformations.py [hunk 28]
```diff
@@ -529,9 +583,217 @@ def fuse_conv_bn(model: torch.fx.GraphModule) -> None:
...
+        model, pattern, replacement, [match_filters]
+    )
+    _compress_qdq_constant_transformation(model, matches_per_channel)
+    _set_meta_for_matches(model, matches_per_channel)
+    pattern, replacement = _get_pattern_replacement_per_tensor()
+    matches_per_tensor = torch.fx.subgraph_rewriter.replace_pattern_with_filters(
+        model, pattern, replacement, [match_filters]
+    )
+    _compress_qdq_constant_transformation(model, matches_per_tensor)
+    _set_meta_for_matches(model.graph, matches_per_tensor)
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch/fx/transformations.py:680]: Let's reduce amount of identical lines
```suggestion
    patterns = [_get_pattern_replacement_per_tensor(), _get_pattern_replacement_per_channel()]
    for pattern, replacement in patterns:
        ...
```

File: nncf/experimental/torch/fx/transformations.py (hunk 25 above)
Comment (daniil-lyakhov) [reply nncf/experimental/torch/fx/transformations.py:790]: Could you please move the all code from `_remove_constant_qdq_transformation` to `compress_post_quantize_transformation`?

**Review by daniil-lyakhov — COMMENTED** [review]
Minor

File: nncf/experimental/torch/fx/transformations.py (hunk 26 above)
Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py:759]: This function is used to fake quantize once so that the rounding error is reduced in OV model. It is used when we are not compressing the weights after quantization, that is, setting backend_params for `compress_weights` to false

File: nncf/experimental/torch/fx/transformations.py (hunk 27 above)
Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py:722]: Alright!

File: nncf/experimental/torch/fx/transformations.py (hunk 28 above)
Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py:680]: Done

File: nncf/experimental/torch/fx/transformations.py (hunk 27 above)
Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py:722]: Done

File: nncf/experimental/torch/fx/transformations.py [hunk 29]
```diff
@@ -529,9 +583,206 @@ def fuse_conv_bn(model: torch.fx.GraphModule) -> None:
...
+        sub_node = match.replacements[1]
+        weight_node, scale_node, zp_node, axis = None, None, None, None
+        nodes_map = {node.name: match.nodes_map[node] for node in match.nodes_map}
+        get_const = partial(get_tensor_constant_from_node, model=model)
+        weight_node = get_const(nodes_map["weight"])
+        scale_node = get_const(nodes_map["scale"])
+        zp_node = get_const(nodes_map["zero_point"])
+        axis = nodes_map["axis"]
+        port_id = 0
+        result = None
```
Comment (daniil-lyakhov) [inline nncf/experimental/torch/fx/transformations.py:680]:
```suggestion
```

**Review by daniil-lyakhov — APPROVED** [review]

File: nncf/experimental/torch/fx/transformations.py (hunk 29 above)
Comment (anzr299) [reply nncf/experimental/torch/fx/transformations.py:680]: Done!

**Review by alexsu52 — APPROVED** [review]
LGTM
